packages/core/src/contrib/Constraints.ts

```diff
@@ -229,6 +229,54 @@ export const objDict = {
   nearScalar: (c: any, goal: any) => {
     return squared(sub(constOfIf(c), constOfIf(goal)));
   },
+  /**
+   * Repel the angle between the p1-p0 and p1-p2 away from 0 and 180 degrees. NOTE: angles between 10 and 170 degrees are considered satisfied.
+   */
+  nonDegenerateAngle: (
+    [, p0]: [string, any],
+    [, p1]: [string, any],
+    [, p2]: [string, any],
+    strength = 20
+  ) => {
+    if (every([p0, p1, p2].map((props) => props["center"]))) {
+      const c0 = fns.center(p0);
+      const c1 = fns.center(p1);
+      const c2 = fns.center(p2);
+
+      const l1 = ops.vsub(c0, c1);
+      const l2 = ops.vsub(c2, c1);
+
+      const cosine = absVal(ops.vdot(ops.vnormalize(l1), ops.vnormalize(l2)));
+      // angles between 10 and 170 degrees do not need to be pushed
+      return ifCond(
+        lt(cosine, varOf(0.985)), // cos10 deg = .9848, abs(cos170) = .9848
+        varOf(0),
+        mul(constOfIf(strength), cosine)
+      );
+    } else {
+      throw new Error(
+        "nonDegenerateAngle: all input shapes need to have centers"
+      );
+    }
+  },
+  /**
+   * try to make distance between a point and a segment `s1` = padding.
+   */
+  pointAtDistFromClosestPt: (
```

Comment:
maybe `pointLineDist` or something similar would be a better name? point-line distance is canonically the shortest distance from the point to the line